pxr/usd/usd/stageCache.cpp

```diff
@@ -65,69 +58,153 @@ std::atomic_long idCounter(9223000);
 Id GetNextId() { return Id::FromLongInt(++idCounter); }
 
 struct Entry {
-    Entry() {}
+    Entry() = default;
     Entry(const UsdStageRefPtr &stage, Id id) : stage(stage), id(id) {}
     UsdStageRefPtr stage;
     Id id;
 };
 
-struct ById {};
+using StagesById = std::unordered_map<Id, UsdStageRefPtr, TfHash>;
+using IdsByStage = std::unordered_map<UsdStageRefPtr, Id, TfHash>;
+using StagesByRootLayer = std::unordered_multimap<SdfLayerHandle,
+                                                  UsdStageRefPtr,
+                                                  TfHash>;
+
+/// The stage container provides a set of synchronization maps to enable
+/// lookup by Id, LayerHandle, or Stage. It provides a single insertion
+/// function to populate these maps and erasure by multiple key types.
+///
+/// This data structure does not add any thread safety guarantees. Users should
+/// lock appropriately before read and write.
+class StageContainer {
+public:
+    StageContainer() = default;
+
+    /// Return an unordered_map of the stages indexed by id
+    const StagesById& ById() const { return _byId; }
+    /// Return an unordered_map of the ids indexed by stage
+    const IdsByStage& ByStage() const { return _byStage; }
+    /// Return an unordered_multimap of the stages indexed by root layer
+    /// NOTE: This is a multimap where multiple stages may mapped by a single
+    /// root layer.
+    const StagesByRootLayer& ByRootLayer() const { return _byRootLayer; }
+
+    // If the stage is already in the cache, return the pair {Id, false}
+    // If the stage is not in the cache, generate a new id, populate
+    // the synchronized maps, and return {NewId, true}
+    std::pair<Id, bool> Insert(const UsdStageRefPtr& stage) {
+        const auto it = _byStage.find(stage);
+        if (  it != _byStage.cend()) {
+            return std::make_pair(it->second, false);
+        }
+        const Id id = GetNextId();
```

Comment:
Notes about all the uses of TF_AXIOM:
1. We definitely do not want TF_AXIOM wrapping any statements with a side effect like the emplace statements here. The docs for TF_AXIOM state that these statements are allowed to be converted to no-ops in an optimized build (even though we don't do that in any builds currently)
2. We prefer to use TF_VERIFY around an expression that is required to return true or even a normal if conditional that produces a TF_CODING_ERROR it the condition is not true.
3. Do we need axioms/verifies here at all?

Reply:
`emplace` returning false would be a red flag that the class's invariants have been violated. I think some sort of unlikely failure assertion would be helpful. I can switch this to a `TF_VERIFY`.